lib/MusicBrainz/Server/Controller/WS/2/Genre.pm

```diff
@@ -0,0 +1,70 @@
+package MusicBrainz::Server::Controller::WS::2::Genre;
+use Moose;
+use MusicBrainz::Server::Validation qw( is_guid );
+
+BEGIN { extends 'MusicBrainz::Server::ControllerBase::WS::2' }
+
+use aliased 'MusicBrainz::Server::WebService::WebServiceStash';
+
+my $ws_defs = Data::OptList::mkopt([
+    genre => {
+        method   => 'GET',
+        required => [ qw(query) ],
+        optional => [ qw(fmt limit offset) ],
+    },
+    genre => {
+        method   => 'GET',
+        optional => [ qw(fmt limit offset) ],
+    },
+    genre => {
+        method   => 'GET',
+        optional => [ qw(fmt) ],
+    }
+]);
+
+with 'MusicBrainz::Server::WebService::Validator' => {
+    defs => $ws_defs,
+};
+
+with 'MusicBrainz::Server::Controller::WS::2::Role::Lookup' => {
+    model => 'Genre',
+};
+
+sub base : Chained('root') PathPart('genre') CaptureArgs(0) { }
+
+sub genre_toplevel {
+    my ($self, $c, $stash, $genres) = @_;
+
+}
+
+sub genre_browse : Private {
+    my ($self, $c) = @_;
+
+    my ($resource, $id) = @{ $c->stash->{linked} };
+    my ($limit, $offset) = $self->_limit_and_offset($c);
+
+    if (!is_guid($id)) {
+        $c->stash->{error} = "Invalid mbid.";
+        $c->detach('bad_req');
+    }
+
+    my $genres;
+
+    my $stash = WebServiceStash->new;
+
+    $self->genre_toplevel($c, $stash, $genres->{items});
+
+    $c->res->content_type($c->stash->{serializer}->mime_type . '; charset=utf-8');
+    $c->res->body($c->stash->{serializer}->serialize('genre-list', $genres, $c->stash->{inc}, $stash));
+}
+
+sub genre_search : Chained('root') PathPart('genre') Args(0) {
+    my ($self, $c) = @_;
+
+    $c->detach('genre_browse') if $c->stash->{linked};
+    $self->_search($c, 'genre');
+}
+
+__PACKAGE__->meta->make_immutable;
+1;
+
```

Comment:
You may want to add our ending boilerplate from [Perl module template](https://musicbrainz.org/doc/Development/Server_Coding_Style#Perl_Module_Template_and_Documentation_Guidelines)